File: algorithms/medium/range_sum_query_2d_immutable.ex
defmodule NumMatrix do
  use GenServer

  @spec init_(matrix :: [[integer]]) :: any
  def init_(matrix) do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)

    matrix =
      matrix
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {row, i}, acc ->
        row
        |> Enum.with_index()
        |> Enum.reduce(acc, fn {n, j}, acc ->
          Map.put(
            acc,
            {i + 1, j + 1},
            Map.get(acc, {i, j + 1}, 0) + Map.get(acc, {i + 1, j}, 0) -
              Map.get(acc, {i, j}, 0) + n
          )
        end)
      end)

    GenServer.call(__MODULE__, {:init, matrix})
  end

  @spec sum_region(row1 :: integer, col1 :: integer, row2 :: integer, col2 :: integer) :: integer
  def sum_region(row1, col1, row2, col2) do
    GenServer.call(__MODULE__, {:sum_region, row1, col1, row2, col2})
  end

  def init(state) do
    {:ok, state}
  end

  def handle_call({:init, mat}, _from, _state) do
    {:reply, true, mat}
  end

  def handle_call({:sum_region, r1, c1, r2, c2}, _from, state) do
    {:reply,
     Map.get(state, {r2 + 1, c2 + 1}, 0) - Map.get(state, {r2 + 1, c1}, 0) -
       Map.get(state, {r1, c2 + 1}, 0) + Map.get(state, {r1, c1}, 0), state}
  end
end

# Your functions will be called as such:
# NumMatrix.init_(matrix)
# param_1 = NumMatrix.sum_region(row1, col1, row2, col2)

# NumMatrix.init_ will be called before every test case, in which you can do some necessary initializations.

File: algorithms/medium/range_sum_query_2d_immutable.rs
struct NumMatrix {
    mat: Vec<Vec<i32>>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl NumMatrix {

    fn new(matrix: Vec<Vec<i32>>) -> Self {
        let mut mat = vec![vec![0; matrix[0].len() + 1]; matrix.len() + 1];
        for i in 1..mat.len() {
            for j in 1..mat[0].len() {
                mat[i][j] = mat[i-1][j] + mat[i][j-1] - mat[i-1][j-1] + matrix[i-1][j-1];
            }
        }
        Self { mat }
    }
    
    fn sum_region(&self, row1: i32, col1: i32, row2: i32, col2: i32) -> i32 {
        let r1 = row1 as usize;
        let c1 = col1 as usize;
        let r2 = row2 as usize;
        let c2 = col2 as usize;
        self.mat[r2+1][c2+1] - self.mat[r2+1][c1] - self.mat[r1][c2+1] + self.mat[r1][c1]
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * let obj = NumMatrix::new(matrix);
 * let ret_1: i32 = obj.sum_region(row1, col1, row2, col2);
 */
